File: apps/web/lib/api/interview-appointment.ts
import api from "@/lib/api"
import type { AppointmentScheduling } from "@/lib/api/appointments"

export type InterviewAppointment = {
    id: string
    scheduled_start: string
    scheduled_end: string
    client_timezone: string
    status: string
    meeting_started_at: string | null
    meeting_ended_at: string | null
    scheduling?: AppointmentScheduling | null
}

export type AppointmentStage = { id: string; label: string; color: string }

export type InterviewAppointmentExternalSyncStatus =
    | "pending"
    | "completed"
    | "failed"
    | "conflict"
    | "unlinked"
    | null

export type InterviewAppointmentState = {
    appointment: InterviewAppointment | null
    can_manage: boolean
    scheduled_stage: AppointmentStage | null
    reschedule_stage: AppointmentStage | null
    external_sync_status: InterviewAppointmentExternalSyncStatus
}

export type InterviewSlots = {
    timezone: string
    slots: { start: string; end: string }[]
}

export type ManageInterviewAppointmentPayload = {
    action: "schedule" | "reschedule" | "cancel"
    scheduled_start?: string
    move_stage: boolean
    expected_stage_id: string
    expected_appointment_id: string | null
    expected_scheduled_start: string | null
    expected_revision?: number
    request_id?: string
    override_availability?: boolean
    override_reason?: string
}

export const getInterviewAppointment = (surrogateId: string) =>
    api.get<InterviewAppointmentState>(`/surrogates/${surrogateId}/interview-appointment`)

export const getInterviewSlots = (surrogateId: string, date: string, timezone: string) =>
    api.get<InterviewSlots>(`/surrogates/${surrogateId}/interview-appointment/slots?date=${encodeURIComponent(date)}&client_timezone=${encodeURIComponent(timezone)}`)

export const manageInterviewAppointment = (surrogateId: string, payload: ManageInterviewAppointmentPayload) =>
    api.post<InterviewAppointmentState>(`/surrogates/${surrogateId}/interview-appointment`, payload)

export const retryInterviewAppointmentGoogleSync = (
    surrogateId: string,
    expectedAppointmentId: string,
) => api.post<InterviewAppointmentState>(
    `/surrogates/${surrogateId}/interview-appointment/sync/retry`,
    { expected_appointment_id: expectedAppointmentId },
)

File: apps/api/app/schemas/interview_appointment.py
"""Contracts for managing a surrogate's initial interview appointment."""

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, model_validator


class InterviewAppointmentRead(BaseModel):
    id: UUID
    scheduled_start: datetime
    scheduled_end: datetime
    client_timezone: str
    status: str
    meeting_started_at: datetime | None
    meeting_ended_at: datetime | None


class InterviewStageRead(BaseModel):
    id: UUID
    label: str
    color: str


class SurrogateInterviewAppointmentState(BaseModel):
    appointment: InterviewAppointmentRead | None
    can_manage: bool
    external_sync_status: (
        Literal["pending", "completed", "failed", "conflict", "unlinked"] | None
    ) = None
    scheduled_stage: InterviewStageRead | None
    reschedule_stage: InterviewStageRead | None


class InterviewGoogleSyncCheck(BaseModel):
    expected_appointment_id: UUID


class SurrogateInterviewAppointmentAction(BaseModel):
    action: Literal["schedule", "reschedule", "cancel"]
    scheduled_start: datetime | None = None
    move_stage: bool
    expected_stage_id: UUID
    expected_appointment_id: UUID | None
    expected_scheduled_start: datetime | None

    @model_validator(mode="after")
    def validate_action(self):
        if self.action in {"schedule", "reschedule"}:
            if self.scheduled_start is None:
                raise ValueError("scheduled_start is required")
            if self.scheduled_start.utcoffset() is None:
                raise ValueError("scheduled_start must include a timezone")
        elif self.scheduled_start is not None:
            raise ValueError("scheduled_start is not allowed when cancelling")
        if (
            self.expected_scheduled_start is not None
            and self.expected_scheduled_start.utcoffset() is None
        ):
            raise ValueError("expected_scheduled_start must include a timezone")
        return self
